Use effective_solar_pathlength_corrected for cloud type and cloud phase

gerritholl opened the pull request:
Use the pure Li/Shibata `effective_solar_pathlength_corrected` modifier for cloud type and cloud phase RGBs for FCI.  This makes the composite usable closer to the terminator, but without the saturation seen in the ABI versions.  The pure correction is achieved by setting the correction limit very high (150 degree).

Add an imagery test and documentation on how to add imagery tests.

See below (also the comments) for different versions.

<!-- Describe what your PR does, and why -->
<!-- For works in progress choose "Create draft pull request" from the drop-down green button. -->

 - [x] Closes #3096<!-- remove if there is no corresponding issue, which should only be the case for minor changes -->
 - [x] Tests added <!-- for all bug fixes or enhancements -->
 - [ ] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->

Cloud type with only `sunz_corrected` (satpy name: `cimss_cloud_type`):

![MTG-I1-fci-eurol-cimss_cloud_type-20250116080000-20250116081000](https://github.com/user-attachments/assets/a054655e-d782-492a-9b05-965f89b8f2c2)

Cloud type with `sunz_corrected` and `sunz_reduced` (satpy name: `cloud_type`):

![MTG-I1-fci-eurol-cloud_type-20250116080000-20250116081000](https://github.com/user-attachments/assets/2bd48e9f-a2b1-4ea9-9933-2db3c918957c)

Cloud type with `solar_pathlength_corrected`:

![MTG-I1-fci-eurol-dwd_cloud_type-20250116080000-20250116081000](https://github.com/user-attachments/assets/03c7a48e-b521-4192-aeb3-b6f5e66ab4e9)

Cloud type with `solar_pathlength_corrected` and `rayleigh_corrected` (for VIS 0.6):

![MTG-I1-fci-eurol-dwd_cloud_type_with_rayleigh-20250116080000-20250116081000](https://github.com/user-attachments/assets/414111eb-f2dc-49a8-b2fe-981afa97bec1)

Cloud phase with `sunz_corrected` and `sunz_reduced`:

![MTG-I1-fci-eurol-cloud_phase-20250116080000-20250116081000](https://github.com/user-attachments/assets/f1ced048-261a-4878-a9c3-0e434fdd5a9c)

Cloud phase with `solar_pathlength_corrected`:

![MTG-I1-fci-eurol-dwd_cloud_phase-20250116080000-20250116081000](https://github.com/user-attachments/assets/5281165b-7458-4394-80df-1a53e52454b9)

Cloud phase with `solar_pathlength_corrected` and `rayleigh_corrected` (for VIS 0.6):

![MTG-I1-fci-eurol-dwd_cloud_phase_with_rayleigh-20250116080000-20250116081000](https://github.com/user-attachments/assets/96c81108-1b6e-4e5f-9eb0-ad343272de91)

gerritholl committed:
Use solar_pathlength_corrected for cloud type and cloud phase

Use the solar_pathlength_corrected modifier for cloud type and cloud
phase RGBs for FCI.  This makes the composite usable closer to the terminator
but without the saturation seen in the ABI versions.

strandgren commented:
Nice work on this, it's clear that there is quite some more information to get from the imagery. But I have two questions:
- in the recipes you use `solar_pathlength_corrected`, but where is this defined? I can only find `effective_solar_pathlength_corrected` defined here https://github.com/pytroll/satpy/blob/main/satpy/etc/composites/visir.yaml#L14.
- Could you also provide a couple of example of how your proposed changes look in the day-night blends, i.e. `cloud_type_with_night_ir105` and `cloud_phase_with_night_ir105`? We may need to tweak the blending there as well.

And as a comment, I still find that at least for the thick part of the frontal cloud close to the terminator the level of saturation is a bit too high. But I'm also aware that it's hard to find settings that works well for all conditions.

gerritholl committed:
solar pathlength -> effective solar pathlength

Fixed modifier name for solar pathlength corrected.

gerritholl commented:
> * in the recipes you use `solar_pathlength_corrected`, but where is this defined?

Oops.  Only in my local recipes.  Fixed with https://github.com/pytroll/satpy/pull/3032/commits/6d5fe19d1e166dc6d001f68a69bbceb972bf6045.  It shows why we need the automated imagery tests, and probably we should indeed merge https://github.com/pytroll/satpy/pull/3013 first (where I added more on the behave tests) so I can leverage this!

Will check the versions with `night_ir105`.

gerritholl commented:
Cloud type night IR 10.5, before:

![MTG-I1-fci-eurol-cloud_type_with_night_ir105-20250116080000-20250116081000](https://github.com/user-attachments/assets/1130d876-8fec-4ac3-80cb-4204ad52b93b)

Cloud type night IR, after:

![MTG-I1-fci-eurol-dwd_cloud_type_with_night_ir105-20250116080000-20250116081000](https://github.com/user-attachments/assets/739ec9de-09a7-4e04-9cd9-34a9a7f218f9)

Cloud phase night IR, before:

![MTG-I1-fci-eurol-cloud_phase_with_night_ir105-20250116080000-20250116081000](https://github.com/user-attachments/assets/6fc72e9f-c160-408f-b09a-932479cd3de0)

Cloud phase night IR, after:

![MTG-I1-fci-eurol-dwd_cloud_phase_with_night_ir105-20250116080000-20250116081000](https://github.com/user-attachments/assets/b5272a34-53fb-4102-9560-cd2c6f52a974)

For curiosity, our internal `solar_pathlength_corrected`, that I accidentally used in the first commit, is defined as:

```
  solar_pathlength_corrected:
    modifier: !!python/name:satpy.modifiers.EffectiveSolarPathLengthCorrector
    optional_prerequisites:
      - solar_zenith_angle
    max_sza: !!null
```

whereas the satpy one is defined as:

https://github.com/pytroll/satpy/blob/40a62fa86cb3c99c70f57e12bba655a47de3ddf2/satpy/etc/composites/visir.yaml#L14-L17

The version I showed in the initial issue description is with `max_sza: !!null`, which we use internally for all our composites.

Cloud type with the effective solar pathlength corrector, default `max_sza`:

![MTG-I1-fci-eurol-dwd_cloud_type-20250116080000-20250116081000-espc](https://github.com/user-attachments/assets/5565483f-5e94-4683-ae87-12867092101d)

Cloud type with the solar pathlength corrector, no `max_sza`:

![MTG-I1-fci-eurol-dwd_cloud_type-20250116080000-20250116081000-spc](https://github.com/user-attachments/assets/133689d6-1f96-4d75-b739-a033a644ddfa)

strandgren commented:
For the `effective_solar_pathlength_corrected` I don't think the reduction is needed and setting `max_sza: !!null` should be fine (as your plots also confirm). For `sunz_corrected` I thinks it's useful, or even needed, given that the correction breaks around 90 degrees otherwise. You can see a comparison here: https://github.com/pytroll/satpy/issues/2955#issuecomment-2437070856.

So perhaps you can add your local  `effective_solar_pathlength_corrected` definition in fci.yaml?

Thanks for the addition day/night blends. Doesn't seem to be much affected. But I believe that with these changes we can do the day/night blending at higher sunz angles to retain more of the daytime imagery, but that might be for another PR.

ameraner commented:
Thank you for your work and analysis! 
I have mixed feelings about this: while I fully agree that trying to include more illuminated daytime pixels is beneficial, both RGBs do not really perform well in that twilight area... In the Cloud Type, the red-toned noise and ESL signal from the NIR1.3 is quite enhanced, and in the Cloud Phase the entire area is dominated by the blue component, making it not too useful. 
It seems however that applying the default `max_sza` removes some of the low-quality and not-so-usable area, while still including a good amout of extra daytime pixels. So maybe that's a good compromise? If I understand the discussion correctly, the current PR code modifications reflects this setup.

gerritholl commented:
There is a tradeoff between æsthetics and the ability to extract information.  This is probably true for any RGB based on solar channels and also when considering the day/night-transition in geo colour.  And any RGB, as well as quantitative products, will lose quality approaching the terminator.  Maybe cloud type is more affected here because NIR 1.3 is so dark already.

For our operational users, the ability to extract information is more important than æsthetics.  I think they will accept and understand that.  Colours are a bit different and noise becomes visible. OK.

> If I understand the discussion correctly, the current PR code modifications reflects this setup.

Yes, the current PR code modification uses the default `solar_pathlength_corrected` (`correction_limit=88`, `max_sza=95`).  It's a compromise between the status quo (`sunz_corrected` with or without `sunz_reducer`) and the DWD version (`correction_limit=88`, `max_sza=None`).  The compromise looks prettier near the terminator, but the DWD version has more extractable information.  What do we want as default in satpy?

ameraner commented:
I agree that information content is more important than aesthetics, at least for the default recipe (a user interested in only PR-like nice looking images can still tune a different recipe to their wishes, but I'd say that the general focus should be on meteorological usability and usefulness). 
My question is if the low twilight area, with all the artefacts like noise and non-representative color tones (that are at that point not describing the actual physical status of the surface/cloud scene, like the violet terminator of the cloud phase and the saturated red clouds in cloud type) is useful/delivering information to the observer. Admittedly, it likely still delivers information at least for spatial patterns, even if the spectral information is likely not very usable... so I don't have a strong opinion, I'm ok with both solutions in the end. 
But at least we can leave a note in the yaml saying that using the default `max_sza` can help to remove some of the bad quality areas, if any user investigates this in the future.

mraspaud commented:
Really nice work!
The discussion re whether to remove the terminator artefacts or not is interesting. Lets no forget though the regions where twilight lasts all day :) for these, I’m suspecting that having data with artefacts can still be better than no data at all...

gerritholl commented:
> The discussion re whether to remove the terminator artefacts or not is interesting.

I'm all for removing artefacts, but removing the entire area that has artefacts in it has something of throwing out the child with the dirty bathwater...

gerritholl committed:
Merge branch 'main' into fci-cloud-type-phase-solar-pathlength-corrector

gerritholl committed:
Merge branch 'main' into fci-cloud-type-phase-solar-pathlength-corrector

gerritholl commented:
See also https://github.com/pytroll/satpy/issues/3096 where @strandgren has made an excellent write-up of the different correction reductions near the terminator.

gerritholl committed:
Merge branch 'main' into fci-cloud-type-phase-solar-pathlength-corrector

solving one merge conflict

gerritholl commented:
Sticking with `EffectiveSolarPathlengthCorrector` with `max_sza: !!null`, varying only `correction_limit` (no Rayleigh correction applied to any channel).

correction limit 86°:

![MTG-I1-fci-eurol-cloud_type_86-20250116080000-20250116081000](https://github.com/user-attachments/assets/79f3502d-0048-4a2f-a71f-e8dd059fdb1c)

correction limit 87°:

![MTG-I1-fci-eurol-cloud_type_87-20250116080000-20250116081000](https://github.com/user-attachments/assets/b7c7d8b2-a0fd-4205-aca3-bee6698905ed)

correction limit 88°:

![MTG-I1-fci-eurol-cloud_type_88-20250116080000-20250116081000](https://github.com/user-attachments/assets/95f025dd-55c8-476d-88a8-de3077e32a89)

correction limit 89°:

![MTG-I1-fci-eurol-cloud_type_89-20250116080000-20250116081000](https://github.com/user-attachments/assets/c4949579-d09c-46ac-8bec-1f487e93371d)

correction limit 90°:

![MTG-I1-fci-eurol-cloud_type_90-20250116080000-20250116081000](https://github.com/user-attachments/assets/b20294b7-1653-4677-9eda-7972e92fc692)

with `night_ir105`:

correction limit 86°:

![MTG-I1-fci-eurol-cloud_type_with_night_ir105_86-20250116080000-20250116081000](https://github.com/user-attachments/assets/f71915ff-30a4-4540-86d5-eb25056a49b7)

correction limit 87°:

![MTG-I1-fci-eurol-cloud_type_with_night_ir105_87-20250116080000-20250116081000](https://github.com/user-attachments/assets/1131dd40-b06d-45a5-bcd9-4a0c6e7552a1)

differences start to become small here, because the right part is taking over when using `lim_low: 78` and `lim_high: 88`, but of course this could be set tighter.

The same for cloud phase:

correction limit 86°:

![MTG-I1-fci-eurol-cloud_phase_86_norayl-20250116080000-20250116081000](https://github.com/user-attachments/assets/8ec4ce91-ebf7-4779-9e56-be1898041813)

correction limit 87°:

![MTG-I1-fci-eurol-cloud_phase_87_norayl-20250116080000-20250116081000](https://github.com/user-attachments/assets/26aa59cb-ccd2-4a34-950b-26b9efc60ed9)

correction limit 88°:

![MTG-I1-fci-eurol-cloud_phase_88_norayl-20250116080000-20250116081000](https://github.com/user-attachments/assets/a7d08e82-b998-49c2-95d9-df93e790594b)

strandgren commented:
@gerritholl Can you please add an example with this configuration as well:
```
correction_limit: 150
max_sza: null
```
which effectively disables the reduction functionality starting at `correction_limit` and thus falls back on the raw Li and Shibata parameterization for the full sunz range.

Sorry for my unavailability this week, ping me on slack if you want me join any discussion on this.

gerritholl commented:
This is with 150°:

![MTG-I1-fci-eurol-cloud_type_150-20250116080000-20250116081000](https://github.com/user-attachments/assets/8c27e77c-b7a0-488a-9e0a-56d17efcd248)

Seems to be less visible near the terminator compared to 90°.

strandgren commented:
> This is with 150°:
> 
> ![MTG-I1-fci-eurol-cloud_type_150-20250116080000-20250116081000](https://private-user-images.githubusercontent.com/500246/443717880-8c27e77c-b7a0-488a-9e0a-56d17efcd248.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDcyMzc1ODAsIm5iZiI6MTc0NzIzNzI4MCwicGF0aCI6Ii81MDAyNDYvNDQzNzE3ODgwLThjMjdlNzdjLWI3YTAtNDg4YS05ZTBhLTU2ZDE3ZWZjZDI0OC5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUxNFQxNTQxMjBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00YTIxZjcxYWU4ZTI3YzdkNTkyZWM4ZWUxOWM1ODBmZTcxMmYwZDAwMjYwMmZiYjRhYzFlMzdkNjkyMzk4MWRjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.3vcQn6GmEGgfTjeepqEWcvEL-_G4aG9JEyBxC4o0VOw)
> 
> Seems to be less visible near the terminator compared to 90°.

Thanks, actually I prefer this one. With 90 degrees there is a red tint for the space part. Furthermore, the ice clouds in the north are less saturated with 150 degrees compared to 90 degrees. I also thinks it an added value if we can say that we use the Li and Shibata parameterization as is, rather than a tweaked version of it.

gerritholl committed:
Set correction limit to 90 and add image test

Set the default correction limit to 90° for the
``effective_solar_pathlength_corrected`` modifier in ``visir.yaml``.

Add image comparison test.

Add documentation on how to add a new image comparison test.

gerritholl committed:
Use raw Li/Shibata by setting correction limit to 150°

mraspaud commented on satpy/etc/composites/fci.yaml:
```diff
@@ -355,18 +355,18 @@ composites:
+        modifiers: [ effective_solar_pathlength_corrected ]
       - name: vis_06
-        modifiers: [sunz_corrected, rayleigh_corrected, sunz_reduced]
+        modifiers: [ effective_solar_pathlength_corrected, rayleigh_corrected ]
```
So at the RGB workshop, in was said that Rayleigh correction for 0.6 µm was optional, except for true color where it was recommended IIRC.
Now I don't have any particular opinion myself on if it needs to be applied or not for cloud phase and cloud type, but it would be nice to be consistent unless there is a specific reason to have it or not.
Not applying the correction could also improve performance.

gerritholl replied:
Good point.  I prefer without, because I think it introduces an artefact.  Removed in abd4f69bb.

mraspaud reviewed (commented):
LGTM, just one question regarding consistency.

gerritholl committed:
Remove cloud phase rayleigh correction

gerritholl reviewed (commented):
start behave test

strandgren commented on satpy/etc/composites/visir.yaml:
```diff
@@ -15,6 +15,9 @@ modifiers:
     modifier: !!python/name:satpy.modifiers.EffectiveSolarPathLengthCorrector
     optional_prerequisites:
       - solar_zenith_angle
+    # max_sza None means no reduction beyond correction_limit, see https://github.com/pytroll/satpy/issues/3096
+    max_sza: !!null
+    correction_limit: 150 # use "raw" Li/Shibata without correction
```
This change will impact any composite that uses this default modifier, so not only the FCI Cloud type and Cloud Phase RGBs. Given the discussion during and past the RGB workshop I guess that's fine and even preferred. But in that case I suggest to modify the PR title and description to make this more clear.

alternatively make a local modifier inside `./satpy/etc/composites/fci.yaml` and then we change the default in a separate PR, potentially the one I want to do at some point refactoring the code for the standard and Li and Shibata sunz corrections.

gerritholl reviewed (commented):
start behave test

gerritholl reviewed (commented):
start behave test

gerritholl reviewed (commented):
start behave test

gerritholl reviewed (commented):
start behave test

gerritholl reviewed (commented):
start behave test

gerritholl reviewed (commented):
start behave test

gerritholl reviewed (commented):
start behave test

gerritholl reviewed (commented):
start behave test

gerritholl reviewed (commented):
start behave test

gerritholl reviewed (commented):
start behave test
